src/Compilers/Core/Portable/DocumentationCommentId.cs

```diff
@@ -301,64 +312,113 @@ private static string DecodePropertyName(string name, string language)
+        /// the return type to see if it failed (in the case of an arbitrary symbol) or that it produced an expected
+        /// value (in the case a known symbol type was used).
+        /// </summary>
+        /// <remarks>
+        /// This will always succeed for a <see cref="INamespaceSymbol"/> or <see cref="INamedTypeSymbol"/>.  It may not
+        /// succeed for other symbols.
+        /// </remarks>
+        private sealed class PrefixAndDeclarationGenerator : SymbolVisitor<bool>
         {
             private readonly StringBuilder _builder;
-            private readonly Generator _generator;
+            private readonly DeclarationGenerator _generator;
 
-            public DeclarationGenerator(StringBuilder builder)
+            public PrefixAndDeclarationGenerator(StringBuilder builder)
             {
                 _builder = builder;
-                _generator = new Generator(builder);
+                _generator = new DeclarationGenerator(builder);
             }
 
-            public override void DefaultVisit(ISymbol symbol)
-            {
-                throw new InvalidOperationException("Cannot generated a documentation comment id for symbol.");
-            }
+            /// <summary>
+            /// If we hit anything we don't know about, indicate failure.
+            /// </summary>
+            public override bool DefaultVisit(ISymbol symbol)
+                => false;
 
-            public override void VisitEvent(IEventSymbol symbol)
+            public override bool VisitEvent(IEventSymbol symbol)
             {
                 _builder.Append("E:");
-                _generator.Visit(symbol);
+                return _generator.Visit(symbol);
             }
 
-            public override void VisitField(IFieldSymbol symbol)
+            public override bool VisitField(IFieldSymbol symbol)
             {
                 _builder.Append("F:");
-                _generator.Visit(symbol);
+                return _generator.Visit(symbol);
             }
 
-            public override void VisitProperty(IPropertySymbol symbol)
+            public override bool VisitProperty(IPropertySymbol symbol)
             {
                 _builder.Append("P:");
-                _generator.Visit(symbol);
+                return _generator.Visit(symbol);
             }
 
-            public override void VisitMethod(IMethodSymbol symbol)
+            public override bool VisitMethod(IMethodSymbol symbol)
             {
                 _builder.Append("M:");
-                _generator.Visit(symbol);
+                return _generator.Visit(symbol);
             }
 
-            public override void VisitNamespace(INamespaceSymbol symbol)
+            public override bool VisitNamespace(INamespaceSymbol symbol)
             {
                 _builder.Append("N:");
-                _generator.Visit(symbol);
+                AppendNamespace(symbol, _builder);
```

Comment:
move to helper that cannot fail.  as such, we can guarantee that this always succeds in the return statement (and doc this in the constracts).